### PR TITLE
Assignment 4

### PR DESCRIPTION
**Pull request by StankovskiA**
This PR contains my solutions to the Tasks 6,7,8 and 9 from the Assignment 4

### CODE REVIEW

File: Assignment4/course_materials/notebooks/task06.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+for s, p, o in g:
+  print(s,p,o)
+
+"""**TASK 6.3: Create a new individual of Researcher named "Jane Smith"**"""
+
+# TO DO
+# Visualize the results
+fullName = Literal("Jane Smith")
+janeURI = ns.JaneSmith
+resource = (janeURI, RDF.type, ns.reasearcher)
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task06.py:55]: same as above

File: Assignment4/course_materials/notebooks/task06.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+for s, p, o in g:
+  print(s,p,o)
+
+"""**TASK 6.5: Add UPM as the university where John Smith works**"""
+
+# TO DO
+# Visualize the results
+upmURI = ns.UPM
+resource = (upmURI, RDF.type, ns.university)
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task06.py:84]: same as above

File: Assignment4/course_materials/notebooks/task07.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+g.namespace_manager.bind('vcard', Namespace("http://www.w3.org/2001/vcard-rdf/3.0#"), override=False)
+g.parse(github_storage+"/rdf/example6.rdf", format="xml")
+
+"""**TASK 7.1: List all subclasses of "LivingThing" with RDFLib and SPARQL**"""
+
+# TO DO
+# Visualize the results
+q1 = """SELECT ?subClass
+WHERE {
+  ?subClass <http://www.w3.org/2000/01/rdf-schema#subClassOf> <http://somewhere#LivingThing> .
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task07.py:30]: this does not retrieve all subclasses of subclasses of LivingThing. Hint: use `*`

File: Assignment4/course_materials/notebooks/task07.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+"""
+
+# TO DO
+# Visualize the results
+
+# Q1 filter out subClasses
+q1 = """SELECT ?x ?z
+WHERE {
+  ?x ?z <http://somewhere#Person> .
+  FILTER (?z != <http://www.w3.org/2000/01/rdf-schema#subClassOf>)
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task07.py:46]: Same problem as above. Also, remove the filter

File: Assignment4/course_materials/notebooks/task07.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+"""**TASK 7.3: List all individuals of "Person" or "Animal" and all their properties including their class with RDFLib and SPARQL. You do not need to list the individuals of the subclasses of person**
+
+"""
+
+# TO DO
+# Visualize the results
+q1 = """SELECT ?x ?z ?class
+WHERE {
+  ?x ?property ?class .
+  FILTER (?property != <http://www.w3.org/2000/01/rdf-schema#subClassOf>) .
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task07.py:72]: Remove all filters. Use a UNION.

File: Assignment4/course_materials/notebooks/task07.py [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+# TO DO
+# Visualize the results
+q1 = """SELECT ?fn
+WHERE {
+  ?x ?property ?class .
+  FILTER (?property != <http://www.w3.org/2000/01/rdf-schema#subClassOf>) .
+  FILTER (?class = <http://somewhere#Person>) .
+  ?x <http://xmlns.com/foaf/0.1/knows> ?y .
+  FILTER (?y = <http://somewhere#RockySmith>) .
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task07.py:90]: Remove filters for this query. If you do ?x ?property ?class, you are basically retrieving the full graph. No query will work in a realistic scenario

File: Assignment4/course_materials/notebooks/task07.py [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+"""**Task 7.5: List the entities who know at least two other entities in the graph**"""
+
+# TO DO
+# Visualize the results
+q1 = """SELECT ?x
+WHERE {
+  ?x <http://xmlns.com/foaf/0.1/knows> ?entity1 .
+  ?x <http://xmlns.com/foaf/0.1/knows> ?entity2 .
+  FILTER (?entity1 != ?entity2)
+} GROUP BY ?x HAVING ((COUNT(DISTINCT ?entity1) + COUNT(DISTINCT ?entity2)) >= 2)"""
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task07.py:106]: You don't need the group by!

File: Assignment4/course_materials/notebooks/task07.py [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+
+for r in g.query(q1):
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task07.py:108]: Missing all exercises in RDFLIB

File: Assignment4/course_materials/notebooks/task08.py [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+properties = ["http://www.w3.org/2001/vcard-rdf/3.0#Given", "http://www.w3.org/2001/vcard-rdf/3.0#Family", "http://www.w3.org/2001/vcard-rdf/3.0#EMAIL"]
+
+# Write dynamic query
+q1 = f"""SELECT ?x ?z
+WHERE {{
+  ?x <http://www.w3.org/1999/02/22-rdf-syntax-ns#type> <http://data.org#Person> .
+  ?x ?y ?z .
+  FILTER (?y != <http://www.w3.org/1999/02/22-rdf-syntax-ns#type>)
+  FILTER (?x = <{persons[1]}>) .
+  FILTER (?y = <{properties[2]}>) .
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task08.py:72]: this is not correct, even if it may work. You should use the triples directly

**Review by dgarijo — CHANGES_REQUESTED** [review]
Please address the remaining issues

**Comment by StankovskiA** [comment]
@dgarijo I have addressed some of the issues in the requested changes.

**Comment by dgarijo** [comment]
@StankovskiA you are not following he guidelines shown in class. You are replacing the course materials instead of creating a folder with your name and student id. Please follow the instructions

File: Assignment4/course_materials/notebooks/task06.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  print(s,p,o)
+
+"""**TASK 6.4: Add to the individual JaneSmith the email address, fullName, given and family names**"""
+
+# TO DO
+# Visualize the results
+VCARD = Namespace("http://www.w3.org/2001/vcard-rdf/3.0#")
+resource = (janeURI, VCARD.email, Literal("smith.jane@gmail.com"))
+g.add(resource)
+resource = (janeURI, VCARD.fn, Literal("Jane Smith"))
```
Comment (dgarijo) [inline Assignment4/course_materials/notebooks/task06.py:57]: minor: fn and email have to be captialized, as it's not the same property in non capital letters

File: Assignment4/course_materials/notebooks/task07.py (hunk 7 above)
Comment (dgarijo) [reply Assignment4/course_materials/notebooks/task07.py:138]: you still don't need group by

**Review by dgarijo — CHANGES_REQUESTED** [review]
Exercises are fine, but we can't merge unless you create your folder following the instructions in class

File: Assignment4/course_materials/notebooks/task07.py (hunk 7 above)
Comment (StankovskiA) [reply Assignment4/course_materials/notebooks/task07.py]: I was unable to get the same results without the grouping

**Review by dgarijo — APPROVED** [review]

File: Assignment4/course_materials/notebooks/task07.py (hunk 7 above)
Comment (dgarijo) [reply Assignment4/course_materials/notebooks/task07.py]: That's odd, you may be introducing errors with the group by then.

**Pull request merged**